src/wire/ipv6hop.rs

```diff
@@ -0,0 +1,308 @@
+    }
+
+    /// Consume the packet, returning the underlying buffer.
+    pub fn into_inner(self) -> T {
+        self.buffer
+    }
+
+    /// Return the next header field.
+    #[inline]
+    pub fn next_header(&self) -> Protocol {
+        let data = self.buffer.as_ref();
+        Protocol::from(data[field::NXT_HDR])
+    }
+
+    /// Return the header length field. Length of the Hop-by-Hop Options header in 8-octet units,
+    /// not including the first 8 octets.
+    #[inline]
+    pub fn header_len(&self) -> u8 {
+        let data = self.buffer.as_ref();
+        data[field::LENGTH]
+    }
+}
+
+impl<'a, T: AsRef<[u8]> + ?Sized> Packet<&'a T> {
+    /// Return the option data.
+    #[inline]
+    pub fn data(&self) -> &'a[u8] {
+        let data = self.buffer.as_ref();
+        &data[field::OPTIONS(data[field::LENGTH])]
+    }
+}
+
+impl<T: AsRef<[u8]> + AsMut<[u8]>> Packet<T> {
+    /// Set the next header field.
+    #[inline]
+    pub fn set_next_header(&mut self, value: Protocol) {
+        let data = self.buffer.as_mut();
+        data[field::NXT_HDR] = value.into();
+    }
+
+    /// Set the option data length. Length of the Hop-by-Hop Options header in 8-octet units,
+    /// not including the first 8 octets.
+    #[inline]
+    pub fn set_header_len(&mut self, value: u8) {
+        let data = self.buffer.as_mut();
+        data[field::LENGTH] = value;
+    }
+}
+
+impl<'a, T: AsRef<[u8]> + AsMut<[u8]> + ?Sized> Packet<&'a mut T> {
+    /// Return a mutable pointer to the option data.
+    #[inline]
+    pub fn data_mut(&mut self) -> &mut [u8] {
+        let data = self.buffer.as_mut();
+        let len = data[field::LENGTH];
+        &mut data[field::OPTIONS(len)]
+    }
+}
+
+impl<'a, T: AsRef<[u8]> + ?Sized> fmt::Display for Packet<&'a T> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match Repr::parse(self) {
+            Ok(repr) => write!(f, "{}", repr),
+            Err(err) => {
+                write!(f, "IPv6 Hop-by-Hop Options ({})", err)?;
+                Ok(())
+            }
+        }
+    }
+}
+
+/// A high-level representation of an IPv6 Hop-by-Hop Options header.
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub struct Repr {
+    /// The type of header immediately following the Hop-by-Hop Options header.
+    pub next_header: Protocol,
+    /// Length of the Hop-by-Hop Options header in 8-octet units, not including the first 8 octets.
+    pub length:      u8,
+}
+
+impl Repr {
+    /// Parse an IPv6 Hop-by-Hop Options Header and return a high-level representation.
+    pub fn parse<T>(packet: &Packet<&T>) -> Result<Repr> where T: AsRef<[u8]> + ?Sized {
+        Ok(Repr {
+            next_header: packet.next_header(),
+            length: packet.header_len()
+        })
+    }
+
+    /// Return the length, in bytes, of a header that will be emitted from this high-level
+    /// representation.
+    pub fn buffer_len(&self) -> usize {
+        let data = field::OPTIONS(self.length);
```

Comment:
Couldn't this just be `field::OPTIONS(self.length).end + 2`?

Reply:
sure, i wasn't trying not to hardcode the 2

Reply:
ah, i can just use `field::OPTIONS(self.length).end` as it already contains the `+ 2` offset.